apm-agent-tracer/src/main/java/co/elastic/apm/agent/tracer/reporting/DataWriter.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.agent.tracer.reporting;
+
+import javax.annotation.Nullable;
+
+public interface DataWriter {
```

Comment:
If I'm not mistaking this class is only used by the OpenTelemetry and Micrometer plugins to report metrics.

In my opinion this abstraction is on the wrong level: Users of this API have to effectively build correct JSON for the elastic specific [IntakeV2 protocol](https://github.com/elastic/apm-server/blob/main/docs/spec/v2/metricset.json).

I think the abstraction should rather be something like a `MetricSetBuilder` where:
 * A metricset is a group of metrics which have the same labels
 * The builder allows adding `double` and `long` scalar metric values
 * The builder allows adding `histogram` metric values
 * The builder terminates with a `report()` operation causing the data to be sent
 
 Maybe it would be even better to allow providing `Labels` with each metric value, abstracting away the grouping into metricsets based on labels from API users.
 
 With #3016 we wanted to remove the duplicate metricset serialization logic, which would make introducing this kind of abstraction a lot simpler. Unfortunately we haven't been able to prioritize this yet and likely won't be able to do so.

Reply:
I thought about a even further refactoring, but I did not want to derive too much from the original implementation in a first attempt. I would prefer to see such a refactoring to be incremental.

Reply:
I agree that this needs a lot more refactoring work to arrive at what I suggested with the `MetricSetBuilder` thingy.

The problem I see with the `DataWrite` is that it introduces an abstraction which actually isn't one. Even with later refactoring, that abstraction doesn't bring any reasonable benefit.

So instead of first getting rid of the core-dependency for the OpenTelemetry and Micrometer instrumentations and then cleaning it up, I think the order should be reversed here to not worsen the code for a potential future refactor, which might be long in the future.

So the order should rather be to first refactor those plugins to allow the introduction of a good abstraction afterwards. I know this means that we'll have to stay with those two plugins depending on `core` for now, as we won't be doing that biggish refactor soon, but for us this is the preferred solution.